### PR TITLE
feat: added hmac option

### PR DESCRIPTION
**Pull request by kierramarie**
### Description

Added the ibm_sm_service_credentials_secret.source_service.parameter HMAC option with corresponding variable. 

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -122,4 +122,10 @@ variable "endpoint_type" {
   }
 }
 
+variable "service_credentials_source_service_hmac" {
+  type        = bool
+  description = "The optional boolean parameter HMAC for creating specific kind of credentials"
```
Comment (alex-reiff) [inline variables.tf:127]: Can this be more explicit? What 'specific kind' or a doc link?

**Review by alex-reiff — CHANGES_REQUESTED** [review]

**Comment by ocofaigh** [comment]
Dont forget to then expose this in https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/tree/main/modules/secrets once merged.
Can you please link the tracking issue in the PR?

**Comment by kierramarie** [comment]
Corresponding issue: https://github.ibm.com/GoldenEye/issues/issues/10804

**Comment by kierramarie** [comment]
Upgrade test failed because the change causes a delete and create of the `ibm_sm_service_credentials_secret`

**Comment by ocofaigh** [comment]
@kierramarie Why would upgrade test fail if we have not enabled HMAC? Can you paste the output?

**Comment by kierramarie** [comment]
@ocofaigh 
```
Should be false
        	Test:       	TestRunUpgradeExample
        	Messages:   	Resource(s) identified to be destroyed 
        	            	Name: service_credentials_secret
        	            	Address: module.secret_manager_service_credential.ibm_sm_service_credentials_secret.service_credentials_secret[0]
        	            	Actions: [delete create]
```
```
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:   # module.secret_manager_service_credential.ibm_sm_service_credentials_secret.service_credentials_secret[0] must be replaced
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185: -/+ resource "ibm_sm_service_credentials_secret" "service_credentials_secret" {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ created_at              = "2024-10-01T14:18:48Z" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ created_by              = "IBMid-666000KAO3" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ credentials             = (sensitive value)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ crn                     = "crn:v1:bluemix:public:secrets-manager:us-south:a/abac0df06b644a9cabc6e44f55b3880e:79c6d411-c18f-4670-b009-b0044a238667:secret:8663a154-c35c-f741-939f-5a0e550e9ef8" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ custom_metadata         = {} -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ downloaded              = true -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ expiration_date         = "2024-12-30T14:18:50Z" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ id                      = "us-south/79c6d411-c18f-4670-b009-b0044a238667/8663a154-c35c-f741-939f-5a0e550e9ef8" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ locks_total             = 0 -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:         name                    = "sm-secret-module-upg-wdje4u-service-credentials"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ next_rotation_date      = "2024-12-29T14:18:48Z" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ secret_id               = "8663a154-c35c-f741-939f-5a0e550e9ef8" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ secret_type             = "service_credentials" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ state                   = 1 -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ state_description       = "active" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ updated_at              = "2024-10-01T14:18:50Z" -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       - version_custom_metadata = {} -> null
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ versions_total          = 1 -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:         # (7 unchanged attributes hidden)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185: 
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:       ~ source_service {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:           ~ iam          = [
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:               - {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                   - apikey    = [
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                       - {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                           - description = "Auto-generated for key crn:v1:bluemix:public:cloud-object-storage:global:a/abac0df06b644a9cabc6e44f55b3880e:ab759ee7-3f9f-4a6a-af04-943d76026509:resource-key:c25b4c0f-c1e3-41ba-aa07-1d58521cf554"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                           - name        = "sm-secret-module-upg-wdje4u-service-credentials"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                         },
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                     ]
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                   - role      = [
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                       - {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                           - crn = "crn:v1:bluemix:public:iam::::serviceRole:Writer"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                         },
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                     ]
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                   - serviceid = [
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                       - {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                           - crn = "crn:v1:bluemix:public:iam-identity::a/abac0df06b644a9cabc6e44f55b3880e::serviceid:ServiceId-1e8a0df6-bebd-4e0a-8524-c00c3c631323"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                         },
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                     ]
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                 },
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:             ] -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:           ~ parameters   = { # forces replacement
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:               + "HMAC" = "false"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:             }
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:           ~ resource_key = [
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:               - {
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                   - crn  = "crn:v1:bluemix:public:cloud-object-storage:global:a/abac0df06b644a9cabc6e44f55b3880e:ab759ee7-3f9f-4a6a-af04-943d76026509:resource-key:c25b4c0f-c1e3-41ba-aa07-1d58521cf554"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                   - name = "sm-secret-module-upg-wdje4u-service-credentials"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:                 },
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:             ] -> (known after apply)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185: 
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:             # (2 unchanged blocks hidden)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:         }
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185: 
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:         # (1 unchanged block hidden)
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:     }
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185: 
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185: Plan: 1 to add, 0 to change, 1 to destroy.
```

**Comment by ocofaigh** [comment]
Oh wow, this is actually a breaking change:

```
~ parameters   = { # forces replacement
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:               + "HMAC" = "false"
TestRunUpgradeExample 2024-10-01T14:19:19Z command.go:185:             }
```

We may need to implement this in a different way. We may need to code it so that if `service_credentials_source_service_hmac` is false, we don't pass any parameters, (as I assume this is the default). And if its set to true, only then pass the value in the `parameters` block.

I also think this should not be a breaking change, so can I ask you to try and reproduce using the provider directly with a simple piece of code using the `ibm_sm_service_credentials_secret` resource, and then create a provider bug with the details. Thanks

**Comment by kierramarie** [comment]
Going to create an issue for creating the provider bug.

**Review by alex-reiff — DISMISSED** [review]

**Comment by kierramarie** [comment]
I have created a provider bug.

File: main.tf [hunk 2]
```diff
@@ -92,8 +92,33 @@ resource "ibm_sm_imported_certificate" "imported_cert" {
   endpoint_type   = var.endpoint_type
 }
 
+# There are two sm cred secrets because adding HMAC parameter causes breaking change. Only one will be deployed.
```
Comment (ocofaigh) [inline main.tf:95]: link to the provider issue

Comment (kierramarie) [reply main.tf:95]: updated

**Review by ocofaigh — COMMENTED** [review]
@kierramarie Did we try to do something like this instead of creating a duplicate `ibm_sm_service_credentials_secret` resource...

`parameters = var.service_credentials_source_service_hmac ? {"HMAC" : var.service_credentials_source_service_hmac} : null`

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**